shaders.go

```diff
@@ -106,193 +107,187 @@ void main (void) {
 	}
-	s.indexVBO = Gl.CreateBuffer()
-	Gl.BindBuffer(Gl.ELEMENT_ARRAY_BUFFER, s.indexVBO)
-	Gl.BufferData(Gl.ELEMENT_ARRAY_BUFFER, s.indices, Gl.STATIC_DRAW)
 
-	s.SetProjection(width, height)
+	if changed := s.generateBufferContent(ren, ren.bufferContent); !changed {
+		return
+	}
 
-	// Define things that should be read from the texture buffer
-	s.inPosition = Gl.GetAttribLocation(s.program, "in_Position")
-	s.inTexCoords = Gl.GetAttribLocation(s.program, "in_TexCoords")
-	s.inColor = Gl.GetAttribLocation(s.program, "in_Color")
+	if ren.buffer == nil {
+		ren.buffer = Gl.CreateBuffer()
+	}
+	Gl.BindBuffer(Gl.ARRAY_BUFFER, ren.buffer)
+	Gl.BufferData(Gl.ARRAY_BUFFER, ren.bufferContent, Gl.STATIC_DRAW)
+}
 
-	// Define things that should be set per draw
-	s.ufPosition = Gl.GetUniformLocation(s.program, "uf_Position")
-	s.ufProjection = Gl.GetUniformLocation(s.program, "uf_Projection")
+func (s *basicShader) generateBufferContent(ren *RenderComponent, buffer []float32) bool {
+	w := ren.Drawable.Width()
+	h := ren.Drawable.Height()
 
-	// Enable those things
-	Gl.EnableVertexAttribArray(s.inPosition)
-	Gl.EnableVertexAttribArray(s.inTexCoords)
-	Gl.EnableVertexAttribArray(s.inColor)
+	colorR, colorG, colorB, colorA := ren.Color.RGBA()
 
-	Gl.Enable(Gl.BLEND)
-	Gl.BlendFunc(Gl.SRC_ALPHA, Gl.ONE_MINUS_SRC_ALPHA)
+	red := colorR
+	green := colorG << 8
+	blue := colorB << 16
+	alpha := colorA << 24
 
-	// TODO: listen for Projection changes
-}
+	tint := math.Float32frombits((alpha | blue | green | red) & 0xfeffffff)
 
-func (s *hudShader) Pre() {
-	Gl.UseProgram(s.program)
-	Gl.Uniform2f(s.ufProjection, s.projX, s.projY)
-}
+	u, v, u2, v2 := ren.Drawable.View()
 
-func (s *hudShader) Draw(texture *gl.Texture, buffer *gl.Buffer, x, y, rotation float32) {
-	if s.lastBuffer != buffer {
-		Gl.BindBuffer(Gl.ARRAY_BUFFER, buffer)
-		Gl.VertexAttribPointer(s.inPosition, 2, Gl.FLOAT, false, 20, 0)
-		Gl.VertexAttribPointer(s.inTexCoords, 2, Gl.FLOAT, false, 20, 8)
-		Gl.VertexAttribPointer(s.inColor, 4, Gl.UNSIGNED_BYTE, true, 20, 16)
+	var changed bool
 
-		s.lastBuffer = buffer
-	}
+	//setValue(buffer, 0, 0, &changed)
+	//setValue(buffer, 1, 0, &changed)
+	setBufferValue(buffer, 2, u, &changed)
+	setBufferValue(buffer, 3, v, &changed)
+	setBufferValue(buffer, 4, tint, &changed)
 
-	if s.lastTexture != texture {
-		Gl.BindTexture(Gl.TEXTURE_2D, texture)
+	setBufferValue(buffer, 5, w, &changed)
+	//setValue(buffer, 6, 0, &changed)
+	setBufferValue(buffer, 7, u2, &changed)
+	setBufferValue(buffer, 8, v, &changed)
+	setBufferValue(buffer, 9, tint, &changed)
 
-		s.lastTexture = texture
-	}
+	setBufferValue(buffer, 10, w, &changed)
+	setBufferValue(buffer, 11, h, &changed)
+	setBufferValue(buffer, 12, u2, &changed)
+	setBufferValue(buffer, 13, v2, &changed)
+	setBufferValue(buffer, 14, tint, &changed)
 
-	Gl.Uniform2f(s.ufPosition, x, y)
-	Gl.DrawElements(Gl.TRIANGLES, 6, Gl.UNSIGNED_SHORT, 0)
-}
+	//setValue(buffer, 15, 0, &changed)
```

Comment:
Lines like these are commented because setting `0` is not needed (they're already 0) - but commented lines to provided the "bigger picture".